### PR TITLE
COLL/TUNED: Update alltoall selection rule for mellanox platform

### PR DESCRIPTION
**Pull request by brminich**
Use linear with sync alltoall algorithm for certain message/comm size ranges. Does not affect default fixed decision, unless HPCX (with its custom parameters) is used or corresponding mca is set.

Signed-off-by: Mikhail Brinskii <mikhailb@mellanox.com>

### CODE REVIEW

**Comment by brminich** [comment]
@yosefe, plz check

File: ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c [hunk 1]
```diff
@@ -115,25 +114,39 @@ int ompi_coll_tuned_alltoall_intra_check_forced_init (coll_tuned_force_algorithm
...
+
     mca_param_indices->max_requests_param_index =
       mca_base_component_var_register(&mca_coll_tuned_component.super.collm_version,
                                       "alltoall_algorithm_max_requests",
                                       "Maximum number of outstanding send or recv requests.  Only has meaning for synchronized algorithms.",
                                       MCA_BASE_VAR_TYPE_INT, NULL, 0, MCA_BASE_VAR_FLAG_SETTABLE,
                                       OPAL_INFO_LVL_5,
                                       MCA_BASE_VAR_SCOPE_ALL,
-                                      &coll_tuned_alltoall_max_requests);
+                                      &ompi_coll_tuned_alltoall_max_reqs);
```
Comment (yosefe) [inline ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c]: maybe keep the old name ("ompi_coll_tuned_alltoall_max_requests")?

File: ompi/mca/coll/tuned/coll_tuned_component.c [hunk 2]
```diff
@@ -57,6 +57,13 @@ int   ompi_coll_tuned_init_max_requests = 128;
 int   ompi_coll_tuned_alltoall_small_msg = 200;
 int   ompi_coll_tuned_alltoall_intermediate_msg = 3000;
 
+/* Set it to intermediate value by default, so it does not affect default
```
Comment (yosefe) [inline ompi/mca/coll/tuned/coll_tuned_component.c]: Set it to same value as intermediate msg by default, ...

File: ompi/mca/coll/tuned/coll_tuned_component.c [hunk 3]
```diff
@@ -57,6 +57,13 @@ int   ompi_coll_tuned_init_max_requests = 128;
...
+ * algorithm selection. Changing this value will force using linear with sync
+ * algorithm on certain message sizes. */
+int   ompi_coll_tuned_alltoall_large_msg = 3000;
+int   ompi_coll_tuned_alltoall_min_procs = 0; /* not used by default */
```
Comment (yosefe) [inline ompi/mca/coll/tuned/coll_tuned_component.c]: /* disable by default */

File: ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c [hunk 4]
```diff
@@ -115,25 +114,39 @@ int ompi_coll_tuned_alltoall_intra_check_forced_init (coll_tuned_force_algorithm
                                       MCA_BASE_VAR_SCOPE_ALL,
                                       &coll_tuned_alltoall_chain_fanout);
 
-    coll_tuned_alltoall_max_requests = 0; /* no limit for alltoall by default */
+    (void) mca_base_component_var_register(&mca_coll_tuned_component.super.collm_version,
+                                           "alltoall_large_msg",
+                                           "threshold (if supported) to decide if large MSGs alltoall algorithm will be used",
```
Comment (yosefe) [inline ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c]: maybe "use pairwise exchange algorithm for communicators larger than this value"

File: ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c [hunk 5]
```diff
@@ -115,25 +114,39 @@ int ompi_coll_tuned_alltoall_intra_check_forced_init (coll_tuned_force_algorithm
...
+                                           MCA_BASE_VAR_TYPE_INT, NULL, 0, 0,
+                                           OPAL_INFO_LVL_6,
+                                           MCA_BASE_VAR_SCOPE_READONLY,
+                                           &ompi_coll_tuned_alltoall_large_msg);
+
+    (void) mca_base_component_var_register(&mca_coll_tuned_component.super.collm_version,
+                                           "alltoall_min_procs",
+                                           "threshold (if supported) to decide if many processes alltoall algorithm will be used",
```
Comment (yosefe) [inline ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c]: maybe "use pairwise exchange algorithm for communicators larger than this value"

**Comment by yosefe** [comment]
@jladd-mlnx @jsquyres can you pls take a look as well?

**Comment by jladd-mlnx** [comment]
@brminich  - how did you determine the new cutoff values for using linear sync? Which platform was used for measurements? HCA/Switch model? If this is based on measurements made on EDR, do we expect them to remain valid for HDR?

**Comment by jladd-mlnx** [comment]
👍 Reviewed data internally; these thresholds are valid for EDR. We may need to reexamine for HDR at a later time.

**Comment by bosilca** [comment]
This is a great starting point to start defining the correct switching points for the Mellanox hardware, and to finally use tuned as it was originally intended to, as a platform for vendors to optimize their own collective communication setup.

**Comment by brminich** [comment]
@yosefe, ok to merge?

**Pull request merged**